firmware/src/adc.c

```diff
@@ -35,12 +43,41 @@ void adc_init() {
 
   // Set ADC1 to continuous mode and overrun mode
   ADC1->CFGR = ADC_CFGR_OVRMOD | ADC_CFGR_CONT;
+
+  // Instruct ADC1 to oversample
+  ADC1->CFGR2 = ADC_CFGR2_ROVSE;
+
+  // Set ADC1 to 256x oversampling, this results in 20 bit total resolution
+  ADC1->CFGR2 |= ADC_CFGR2_OVSR_0 | ADC_CFGR2_OVSR_1 | ADC_CFGR2_OVSR_2;
+
+  // Set ADC1 to 3-bit right shift, this results in 14 bit output, the lowest 3 bits are discarded
+  // Otherwise, at 20 bits of resolution, the output would always be clipped to 65363, because the
+  // DR register is only 16 bits wide
+  ADC1->CFGR2 |= ADC_CFGR2_OVSS_0 | ADC_CFGR2_OVSS_1;
```

Comment:
This is incorrect. You need to right-shift by 4, discarding 4 bits to fit a 20-bit output into a 16 bit register.